## DATA PREPARATION

### GETTING DATA FROM THE SOURCES

In [1]:
# Import requisite libraries
import pyodbc
from dotenv import dotenv_values
from dotenv import load_dotenv
import warnings 
warnings.filterwarnings('ignore')
import os
import requests
from io import StringIO
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

#### DATASET HOSTED AT A GITHUB REPOSITORY

In [2]:
# Github repo url link
url = 'https://raw.githubusercontent.com/Azubi-Africa/Career_Accelerator_LP1-Data_Analysis/main/startup_funding2018.csv'

# Reading data from the Github repo link
data1 = pd.read_csv(url)

data1.tail(5)

Company Name                                          Industry  \
521              Udaan  B2B, Business Development, Internet, Marketplace   
522  Happyeasygo Group                                   Tourism, Travel   
523             Mombay        Food and Beverage, Food Delivery, Internet   
524         Droni Tech                            Information Technology   
525            Netmeds        Biotechnology, Health Care, Pharmaceutical   

    Round/Series       Amount                     Location  \
521     Series C    225000000  Bangalore, Karnataka, India   
522     Series A            —      Haryana, Haryana, India   
523         Seed         7500   Mumbai, Maharashtra, India   
524         Seed  ₹35,000,000   Mumbai, Maharashtra, India   
525     Series C     35000000   Chennai, Tamil Nadu, India   

                                         About Company  
521  Udaan is a B2B trade platform, designed specif...  
522            HappyEasyGo is an online travel domain.  
523  Mombay is a unique opportunity for housewives ...  
524  Droni Tech manufacture UAVs and develop softwa...  
525       Welcome to India's most convenient pharmacy!

In [3]:
#Saving the dataset
data1.to_csv('../Datasets/startup_funding2018.csv', index=False)

#### SQL SERVER DATASETS

In [4]:
# Load environment variables from .env file into a dictionary
environment_variables = dotenv_values("../.env")

# Get the values for the credentials you set in the '.env' file
server    = environment_variables.get("SERVER")
database  = environment_variables.get("DATABASE")
username  = environment_variables.get("USERNAME")
password  = environment_variables.get("PASSWORD")

connection_string = f"DRIVER={{SQL Server}};SERVER={server};DATABASE={database};UID={username};PWD={password};MARS_Connection=yes;MinProtocolVersion=TLSv1.2;"

# Establish a connection to the DAPS Project database using the PyODBC library
cnxn = pyodbc.connect(connection_string)


In [5]:
# Querying data from a DAP database table
query1 = "SELECT * FROM dbo.LP1_startup_funding2020"

data3 = pd.read_sql(query1, cnxn)

data3.tail(5)

Company_Brand  Founded HeadQuarter     Sector  \
1050  Leverage Edu      NaN       Delhi     Edtech   
1051         EpiFi      NaN        None    Fintech   
1052       Purplle   2012.0      Mumbai  Cosmetics   
1053        Shuttl   2015.0       Delhi  Transport   
1054         Pando   2017.0     Chennai   Logitech   

                                           What_it_does  \
1050  AI enabled marketplace that provides career gu...   
1051  It offers customers with a single interface fo...   
1052         Online makeup and beauty products retailer   
1053                    App based bus aggregator serice   
1054            Networked logistics management software   

                             Founders                               Investor  \
1050                Akshay Chaturvedi  DSG Consumer Partners, Blume Ventures   
1051  Sujith Narayanan, Sumit Gwalani          Sequoia India, Ribbit Capital   
1052        Manish Taneja, Rahul Dash                             Verlinvest   
1053    Amit Singh, Deepanshu Malviya             SIG Global India Fund LLP.   
1054   Jayakrishnan, Abhijeet Manohar                      Chiratae Ventures   

          Amount       Stage column10  
1050   1500000.0        None     None  
1051  13200000.0  Seed Round     None  
1052   8000000.0        None     None  
1053   8043000.0    Series C     None  
1054   9000000.0    Series A     None

In [6]:
# Saving the dataframe
data3.to_csv('../Datasets/startup_funding2020.csv', index=False)

In [7]:
# Querying data from another DAP database table
query2 = "SELECT * FROM dbo.LP1_startup_funding2021"

data4 = pd.read_sql(query2, cnxn)

data4.tail(5)

Company_Brand  Founded HeadQuarter                 Sector  \
1204      Gigforce   2019.0    Gurugram  Staffing & Recruiting   
1205        Vahdam   2015.0   New Delhi       Food & Beverages   
1206  Leap Finance   2019.0   Bangalore     Financial Services   
1207  CollegeDekho   2015.0    Gurugram                 EdTech   
1208        WeRize   2019.0   Bangalore     Financial Services   

                                           What_it_does  \
1204                  A gig/on-demand staffing company.   
1205  VAHDAM is among the world’s first vertically i...   
1206  International education loans for high potenti...   
1207  Collegedekho.com is Student’s Partner, Friend ...   
1208  India’s first socially distributed full stack ...   

                           Founders                          Investor  \
1204    Chirag Mittal, Anirudh Syal                   Endiya Partners   
1205                     Bala Sarda                          IIFL AMC   
1206     Arnav Kumar, Vaibhav Singh                      Owl Ventures   
1207                   Ruchir Arora  Winter Capital, ETS, Man Capital   
1208  Vishal Chopra, Himanshu Gupta    3one4 Capital, Kalaari Capital   

         Amount         Stage  
1204   $3000000  Pre-series A  
1205  $20000000      Series D  
1206  $55000000      Series C  
1207  $26000000      Series B  
1208   $8000000      Series A

In [8]:
data4.to_csv('../Datasets/startup_funding2021.csv', index=False)

## LOADING AND EXPLORING THE DATASETS

### LOADING THE DATASETS

In [10]:
# 2018 dataset as df1
df1 = pd.read_csv('../Datasets/startup_funding2018.csv')

df1.head(5)

Company Name                                           Industry  \
0  TheCollegeFever  Brand Marketing, Event Promotion, Marketing, S...   
1  Happy Cow Dairy                               Agriculture, Farming   
2       MyLoanCare   Credit, Financial Services, Lending, Marketplace   
3      PayMe India                        Financial Services, FinTech   
4         Eunimart                 E-Commerce Platforms, Retail, SaaS   

  Round/Series       Amount                          Location  \
0         Seed       250000       Bangalore, Karnataka, India   
1         Seed  ₹40,000,000        Mumbai, Maharashtra, India   
2     Series A  ₹65,000,000           Gurgaon, Haryana, India   
3        Angel      2000000       Noida, Uttar Pradesh, India   
4         Seed            —  Hyderabad, Andhra Pradesh, India   

                                       About Company  
0  TheCollegeFever is a hub for fun, fiesta and f...  
1  A startup which aggregates milk from dairy far...  
2          Leading Online Loans Marketplace in India  
3  PayMe India is an innovative FinTech organizat...  
4  Eunimart is a one stop solution for merchants ...

In [11]:
# 2019 dataset as df2
df2 = pd.read_csv('../Datasets/startup_funding2019.csv')

df2.head(5)

Company/Brand  Founded HeadQuarter           Sector  \
0  Bombay Shaving      NaN         NaN        Ecommerce   
1       Ruangguru   2014.0      Mumbai           Edtech   
2        Eduisfun      NaN      Mumbai           Edtech   
3        HomeLane   2014.0     Chennai  Interior design   
4        Nu Genes   2004.0   Telangana         AgriTech   

                                        What it does  \
0         Provides a range of male grooming products   
1  A learning platform that provides topic-based ...   
2            It aims to make learning fun via games.   
3              Provides interior designing solutions   
4  It is a seed company engaged in production, pr...   

                                Founders  \
0                     Shantanu Deshpande   
1  Adamas Belva Syah Devara, Iman Usman.   
2                          Jatin Solanki   
3           Srikanth Iyer, Rama Harinath   
4                 Narayana Reddy Punyala   

                                            Investor     Amount($)  \
0                               Sixth Sense Ventures    $6,300,000   
1                                   General Atlantic  $150,000,000   
2     Deepak Parekh, Amitabh Bachchan, Piyush Pandey   $28,000,000   
3  Evolvence India Fund (EIF), Pidilite Group, FJ...   $30,000,000   
4           Innovation in Food and Agriculture (IFA)    $6,000,000   

           Stage  
0            NaN  
1       Series C  
2  Fresh funding  
3       Series D  
4            NaN

In [12]:
# 2020 dataset as df3
df3 = pd.read_csv('../Datasets/startup_funding2020.csv')

df3.head(5)

Company_Brand  Founded HeadQuarter              Sector  \
0    Aqgromalin   2019.0     Chennai            AgriTech   
1      Krayonnz   2019.0   Bangalore              EdTech   
2  PadCare Labs   2018.0        Pune  Hygiene management   
3         NCOME   2020.0   New Delhi              Escrow   
4    Gramophone   2016.0      Indore            AgriTech   

                                        What_it_does  \
0                       Cultivating Ideas for Profit   
1  An academy-guardian-scholar centric ecosystem ...   
2   Converting bio-hazardous waste to harmless waste   
3                       Escrow-as-a-service platform   
4  Gramophone is an AgTech platform enabling acce...   

                                            Founders  \
0                    Prasanna Manogaran, Bharani C L   
1                   Saurabh Dixit, Gurudutt Upadhyay   
2                                    Ajinkya Dhariya   
3                                      Ritesh Tiwari   
4  Ashish Rajan Singh, Harshit Gupta, Nishant Mah...   

                              Investor    Amount     Stage column10  
0                      Angel investors  200000.0       NaN      NaN  
1                      GSF Accelerator  100000.0  Pre-seed      NaN  
2                       Venture Center       NaN  Pre-seed      NaN  
3  Venture Catalysts, PointOne Capital  400000.0       NaN      NaN  
4  Siana Capital Management, Info Edge  340000.0       NaN      NaN

In [13]:
# 2021 dataset as df4
df4 = pd.read_csv('../Datasets/startup_funding2021.csv')

df4.head(5)

Company_Brand  Founded HeadQuarter          Sector  \
0  Unbox Robotics   2019.0   Bangalore      AI startup   
1          upGrad   2015.0      Mumbai          EdTech   
2     Lead School   2012.0      Mumbai          EdTech   
3         Bizongo   2015.0      Mumbai  B2B E-commerce   
4        FypMoney   2021.0    Gurugram         FinTech   

                                        What_it_does  \
0  Unbox Robotics builds on-demand AI-driven ware...   
1     UpGrad is an online higher education platform.   
2  LEAD School offers technology based school tra...   
3  Bizongo is a business-to-business online marke...   
4  FypMoney is Digital NEO Bank for Teenagers, em...   

                                            Founders  \
0                        Pramod Ghadge, Shahid Memon   
1  Mayank Kumar, Phalgun Kompalli, Ravijot Chugh,...   
2                         Smita Deorah, Sumeet Mehta   
3            Aniket Deb, Ankit Tomar, Sachin Agrawal   
4                                      Kapil Banwari   

                                        Investor        Amount         Stage  
0                    BEENEXT, Entrepreneur First    $1,200,000  Pre-series A  
1       Unilazer Ventures, IIFL Asset Management  $120,000,000           NaN  
2               GSV Ventures, Westbridge Capital   $30,000,000      Series D  
3                         CDC Group, IDG Capital   $51,000,000      Series C  
4  Liberatha Kallat, Mukesh Yadav, Dinesh Nagpal    $2,000,000          Seed

### EXPLORING THE DATASETS

#### 2018 DATASET

In [14]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 526 entries, 0 to 525
Data columns (total 6 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   Company Name   526 non-null    object
 1   Industry       526 non-null    object
 2   Round/Series   526 non-null    object
 3   Amount         526 non-null    object
 4   Location       526 non-null    object
 5   About Company  526 non-null    object
dtypes: object(6)
memory usage: 24.8+ KB


In [15]:
df1.describe()

Company Name Industry Round/Series Amount  \
count               526      526          526    526   
unique              525      405           21    198   
top     TheCollegeFever        —         Seed      —   
freq                  2       30          280    148   

                           Location  \
count                           526   
unique                           50   
top     Bangalore, Karnataka, India   
freq                            102   

                                            About Company  
count                                                 526  
unique                                                524  
top     TheCollegeFever is a hub for fun, fiesta and f...  
freq                                                    2

In [16]:
# Checking if '—' exists in all the columns
columns_to_check = ['Company Name', 'Industry', 'Round/Series', 'Amount', 'Location', 'About Company']
for column_name in columns_to_check:
    is_in_column = '—' in df1[column_name].values
    if is_in_column:
        print("'—' exists in column", column_name)
    else:
        print("'—' does not exist in column", column_name)

'—' does not exist in column Company Name
'—' exists in column Industry
'—' does not exist in column Round/Series
'—' exists in column Amount
'—' does not exist in column Location
'—' does not exist in column About Company


In [17]:
# Checking the number of occurrences of the '—' character in columns where it exists
print(df1.Industry.value_counts()['—'])

print('\n')

print(df1.Amount.value_counts()['—'])

30


148


#### 

#### 2019 DATASET

In [18]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 89 entries, 0 to 88
Data columns (total 9 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Company/Brand  89 non-null     object 
 1   Founded        60 non-null     float64
 2   HeadQuarter    70 non-null     object 
 3   Sector         84 non-null     object 
 4   What it does   89 non-null     object 
 5   Founders       86 non-null     object 
 6   Investor       89 non-null     object 
 7   Amount($)      89 non-null     object 
 8   Stage          43 non-null     object 
dtypes: float64(1), object(8)
memory usage: 6.4+ KB


In [19]:
df2.describe()

Founded
count    60.000000
mean   2014.533333
std       2.937003
min    2004.000000
25%    2013.000000
50%    2015.000000
75%    2016.250000
max    2019.000000

#### 2020 DATASET

In [20]:
df3.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1055 entries, 0 to 1054
Data columns (total 10 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Company_Brand  1055 non-null   object 
 1   Founded        842 non-null    float64
 2   HeadQuarter    961 non-null    object 
 3   Sector         1042 non-null   object 
 4   What_it_does   1055 non-null   object 
 5   Founders       1043 non-null   object 
 6   Investor       1017 non-null   object 
 7   Amount         801 non-null    float64
 8   Stage          591 non-null    object 
 9   column10       2 non-null      object 
dtypes: float64(2), object(8)
memory usage: 82.6+ KB


In [21]:
df3.describe()

Founded        Amount
count   842.000000  8.010000e+02
mean   2015.363420  1.130430e+08
std       4.097909  2.476635e+09
min    1973.000000  1.270000e+04
25%    2014.000000  1.000000e+06
50%    2016.000000  3.000000e+06
75%    2018.000000  1.100000e+07
max    2020.000000  7.000000e+10

#### 2021 DATASET

In [22]:
df4.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1209 entries, 0 to 1208
Data columns (total 9 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Company_Brand  1209 non-null   object 
 1   Founded        1208 non-null   float64
 2   HeadQuarter    1208 non-null   object 
 3   Sector         1209 non-null   object 
 4   What_it_does   1209 non-null   object 
 5   Founders       1205 non-null   object 
 6   Investor       1147 non-null   object 
 7   Amount         1206 non-null   object 
 8   Stage          781 non-null    object 
dtypes: float64(1), object(8)
memory usage: 85.1+ KB


In [23]:
df4.describe()

Founded
count  1208.000000
mean   2016.655629
std       4.517364
min    1963.000000
25%    2015.000000
50%    2018.000000
75%    2020.000000
max    2021.000000

### DATA CLEANING

#### 2018 DATASET

In [24]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 526 entries, 0 to 525
Data columns (total 6 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   Company Name   526 non-null    object
 1   Industry       526 non-null    object
 2   Round/Series   526 non-null    object
 3   Amount         526 non-null    object
 4   Location       526 non-null    object
 5   About Company  526 non-null    object
dtypes: object(6)
memory usage: 24.8+ KB


In [25]:
# Add a Year column
year = 2018
df1['Year'] = year

In [26]:
# Create Headquarters column
df1['Headquarters'] = df1.Location.str.split().str.get(0)
df1['Headquarters']

0      Bangalore,
1         Mumbai,
2        Gurgaon,
3          Noida,
4      Hyderabad,
          ...    
521    Bangalore,
522      Haryana,
523       Mumbai,
524       Mumbai,
525      Chennai,
Name: Headquarters, Length: 526, dtype: object

In [27]:
# Creat Sector column
df1['Sector']= df1.Industry.str.split().str.get(-1)
df1['Sector']

0           Ticketing
1             Farming
2         Marketplace
3             FinTech
4                SaaS
            ...      
521       Marketplace
522            Travel
523          Internet
524        Technology
525    Pharmaceutical
Name: Sector, Length: 526, dtype: object

In [28]:
# Rename some columns
df1.rename(columns={'Company Name': 'Company_Name', 'Round/Series': 'Stage', 'About Company': 'About'}, inplace=True)
df1.tail(5)

Company_Name                                          Industry  \
521              Udaan  B2B, Business Development, Internet, Marketplace   
522  Happyeasygo Group                                   Tourism, Travel   
523             Mombay        Food and Beverage, Food Delivery, Internet   
524         Droni Tech                            Information Technology   
525            Netmeds        Biotechnology, Health Care, Pharmaceutical   

        Stage       Amount                     Location  \
521  Series C    225000000  Bangalore, Karnataka, India   
522  Series A            —      Haryana, Haryana, India   
523      Seed         7500   Mumbai, Maharashtra, India   
524      Seed  ₹35,000,000   Mumbai, Maharashtra, India   
525  Series C     35000000   Chennai, Tamil Nadu, India   

                                                 About  Year Headquarters  \
521  Udaan is a B2B trade platform, designed specif...  2018   Bangalore,   
522            HappyEasyGo is an online travel domain.  2018     Haryana,   
523  Mombay is a unique opportunity for housewives ...  2018      Mumbai,   
524  Droni Tech manufacture UAVs and develop softwa...  2018      Mumbai,   
525       Welcome to India's most convenient pharmacy!  2018     Chennai,   

             Sector  
521     Marketplace  
522          Travel  
523        Internet  
524      Technology  
525  Pharmaceutical

In [29]:
# Drop Industry and Location columns
df1 = df1.drop(['Industry', 'Location'], axis=1)
df1.tail(5)

Company_Name     Stage       Amount  \
521              Udaan  Series C    225000000   
522  Happyeasygo Group  Series A            —   
523             Mombay      Seed         7500   
524         Droni Tech      Seed  ₹35,000,000   
525            Netmeds  Series C     35000000   

                                                 About  Year Headquarters  \
521  Udaan is a B2B trade platform, designed specif...  2018   Bangalore,   
522            HappyEasyGo is an online travel domain.  2018     Haryana,   
523  Mombay is a unique opportunity for housewives ...  2018      Mumbai,   
524  Droni Tech manufacture UAVs and develop softwa...  2018      Mumbai,   
525       Welcome to India's most convenient pharmacy!  2018     Chennai,   

             Sector  
521     Marketplace  
522          Travel  
523        Internet  
524      Technology  
525  Pharmaceutical

In [30]:
# View unique values in the Amount column
df1.Amount.unique()

array(['250000', '₹40,000,000', '₹65,000,000', '2000000', '—', '1600000',
       '₹16,000,000', '₹50,000,000', '₹100,000,000', '150000', '1100000',
       '₹500,000', '6000000', '650000', '₹35,000,000', '₹64,000,000',
       '₹20,000,000', '1000000', '5000000', '4000000', '₹30,000,000',
       '2800000', '1700000', '1300000', '₹5,000,000', '₹12,500,000',
       '₹15,000,000', '500000', '₹104,000,000', '₹45,000,000', '13400000',
       '₹25,000,000', '₹26,400,000', '₹8,000,000', '₹60,000', '9000000',
       '100000', '20000', '120000', '₹34,000,000', '₹342,000,000',
       '$143,145', '₹600,000,000', '$742,000,000', '₹1,000,000,000',
       '₹2,000,000,000', '$3,980,000', '$10,000', '₹100,000',
       '₹250,000,000', '$1,000,000,000', '$7,000,000', '$35,000,000',
       '₹550,000,000', '$28,500,000', '$2,000,000', '₹240,000,000',
       '₹120,000,000', '$2,400,000', '$30,000,000', '₹2,500,000,000',
       '$23,000,000', '$150,000', '$11,000,000', '₹44,000,000',
       '$3,240,000', '₹60

In [31]:
# Remove comma from all rows
df1['Amount'] = df1['Amount'].str.replace(',', '')
df1.Amount.head(5)

0       250000
1    ₹40000000
2    ₹65000000
3      2000000
4            —
Name: Amount, dtype: object

In [32]:
# Function to convert rupees to dollars
def convert_to_dollars(Amount):
    if Amount.startswith('₹'):
        return float(Amount[1:]) * 0.0146
    else:
        return Amount
 
# Apply the conversion function to the Amount column
df1['Amount'] = df1['Amount'].apply(convert_to_dollars)
df1.Amount.head(5)

0      250000
1    584000.0
2    949000.0
3     2000000
4           —
Name: Amount, dtype: object

In [33]:
# Another view at outcome of comma removal
df1.Amount.unique()

array(['250000', 584000.0, 949000.0, '2000000', '—', '1600000', 233600.0,
       730000.0, 1460000.0, '150000', '1100000', 7300.0, '6000000',
       '650000', 511000.0, 934400.0, 292000.0, '1000000', '5000000',
       '4000000', 438000.0, '2800000', '1700000', '1300000', 73000.0,
       182500.0, 219000.0, '500000', 1518400.0, 657000.0, '13400000',
       365000.0, 385440.0, 116800.0, 876.0, '9000000', '100000', '20000',
       '120000', 496400.0, 4993200.0, '$143145', 8760000.0, '$742000000',
       14600000.0, 29200000.0, '$3980000', '$10000', 1460.0, 3650000.0,
       '$1000000000', '$7000000', '$35000000', 8030000.0, '$28500000',
       '$2000000', 3504000.0, 1752000.0, '$2400000', '$30000000',
       36500000.0, '$23000000', '$150000', '$11000000', 642400.0,
       '$3240000', 876000.0, '$540000000', 9490000.0, 23360000.0,
       '$900000', '$10000000', '$1500000', 1022000.0, '$1000000',
       '$5000000', '$14000000', 1496500.0, '$100000000', 17520.0,
       75920000.0, '$800000'

In [34]:
# Function to clean the Amount column

def clean_column(Amount):
    # Convert the series to string type
    Amount = Amount.astype(str)
    
    # Remove characters: $, space, .
    Amount = Amount.str.replace('[$ .]', '')

    Amount = Amount.str.replace('$', '')

    # Remove single quotation marks surrounding numbers
    Amount = Amount.str.replace(r"'(\d+)'", r'\1')

    # Replace '—' with 0
    Amount = Amount.str.replace('—', '0')

    return Amount

# Using the clean_column function
df1['Amount'] = clean_column(df1['Amount'])
df1['Amount'].head(5)


0      250000
1    584000.0
2    949000.0
3     2000000
4           0
Name: Amount, dtype: object

In [35]:
# Convert the Amount column to float data type
df1['Amount'] = df1['Amount'].astype(float)
print(df1['Amount'].dtype)

float64


In [36]:
# Fill null values in the Sector column with its mode value
df1['Sector'] = df1['Sector'].fillna(df1['Sector'].mode()[0])


In [37]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 526 entries, 0 to 525
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Company_Name  526 non-null    object 
 1   Stage         526 non-null    object 
 2   Amount        526 non-null    float64
 3   About         526 non-null    object 
 4   Year          526 non-null    int64  
 5   Headquarters  526 non-null    object 
 6   Sector        526 non-null    object 
dtypes: float64(1), int64(1), object(5)
memory usage: 28.9+ KB


In [38]:
df1.describe().T

count          mean           std     min     25%       50%  \
Amount  526.0  1.262695e+07  6.646323e+07     0.0     0.0  584000.0   
Year    526.0  2.018000e+03  0.000000e+00  2018.0  2018.0    2018.0   

              75%           max  
Amount  3000000.0  1.000000e+09  
Year       2018.0  2.018000e+03

In [39]:
df1.duplicated().sum()

1

#### 2019 DATASET

In [40]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 89 entries, 0 to 88
Data columns (total 9 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Company/Brand  89 non-null     object 
 1   Founded        60 non-null     float64
 2   HeadQuarter    70 non-null     object 
 3   Sector         84 non-null     object 
 4   What it does   89 non-null     object 
 5   Founders       86 non-null     object 
 6   Investor       89 non-null     object 
 7   Amount($)      89 non-null     object 
 8   Stage          43 non-null     object 
dtypes: float64(1), object(8)
memory usage: 6.4+ KB


In [41]:
# Add a Year column
year = 2019
df2['Year'] = year

In [42]:
df2.head(5)

Company/Brand  Founded HeadQuarter           Sector  \
0  Bombay Shaving      NaN         NaN        Ecommerce   
1       Ruangguru   2014.0      Mumbai           Edtech   
2        Eduisfun      NaN      Mumbai           Edtech   
3        HomeLane   2014.0     Chennai  Interior design   
4        Nu Genes   2004.0   Telangana         AgriTech   

                                        What it does  \
0         Provides a range of male grooming products   
1  A learning platform that provides topic-based ...   
2            It aims to make learning fun via games.   
3              Provides interior designing solutions   
4  It is a seed company engaged in production, pr...   

                                Founders  \
0                     Shantanu Deshpande   
1  Adamas Belva Syah Devara, Iman Usman.   
2                          Jatin Solanki   
3           Srikanth Iyer, Rama Harinath   
4                 Narayana Reddy Punyala   

                                            Investor     Amount($)  \
0                               Sixth Sense Ventures    $6,300,000   
1                                   General Atlantic  $150,000,000   
2     Deepak Parekh, Amitabh Bachchan, Piyush Pandey   $28,000,000   
3  Evolvence India Fund (EIF), Pidilite Group, FJ...   $30,000,000   
4           Innovation in Food and Agriculture (IFA)    $6,000,000   

           Stage  Year  
0            NaN  2019  
1       Series C  2019  
2  Fresh funding  2019  
3       Series D  2019  
4            NaN  2019

In [43]:
# Check number of unique values in Founded column
df2['Founded'].unique()

array([  nan, 2014., 2004., 2013., 2010., 2018., 2019., 2017., 2011.,
       2015., 2016., 2012., 2008.])

In [44]:
# Create function to fill missing values of the Founded with its mode
def fill_null_with_mode(df, column_name):
    mode_value = df[column_name].mode()  # Calculate the mode of the column
    if not mode_value.empty:
        mode_value = mode_value.iloc[0]  # Get mode value
        df[column_name].fillna(mode_value, inplace=True)  # Fill null values with the mode


fill_null_with_mode(df2, 'Founded')
df2.Founded.isna().sum()

0

In [45]:
# Rename some columns
df2.rename(columns={'Company/Brand': 'Company_Name',
                    'HeadQuarter': 'Headquarters',
                    'What it does': 'About',
                    'Amount($)': 'Amount'},
           inplace=True)
df2.tail(5)

Company_Name  Founded Headquarters                   Sector  \
84     Infra.Market   2015.0       Mumbai                Infratech   
85              Oyo   2013.0     Gurugram              Hospitality   
86       GoMechanic   2016.0        Delhi  Automobile & Technology   
87           Spinny   2015.0        Delhi               Automobile   
88  Ess Kay Fincorp   2015.0    Rajasthan                  Banking   

                                                About  \
84  It connects client requirements to their suppl...   
85                Provides rooms for comfortable stay   
86  Find automobile repair and maintenance service...   
87                                Online car retailer   
88              Organised Non-Banking Finance Company   

                                             Founders  \
84                    Aaditya Sharda, Souvik Sengupta   
85                                     Ritesh Agarwal   
86  Amit Bhasin, Kushal Karwa, Nitin Rana, Rishabh...   
87  Niraj Singh, Ramanshu Mahaur, Ganesh Pawar, Mo...   
88                                     Rajendra Setia   

                                             Investor        Amount     Stage  \
84  Tiger Global, Nexus Venture Partners, Accel Pa...   $20,000,000  Series A   
85  MyPreferred Transformation, Avendus Finance, S...  $693,000,000       NaN   
86                                    Sequoia Capital    $5,000,000  Series B   
87  Norwest Venture Partners, General Catalyst, Fu...   $50,000,000       NaN   
88     TPG, Norwest Venture Partners, Evolvence India   $33,000,000       NaN   

    Year  
84  2019  
85  2019  
86  2019  
87  2019  
88  2019

In [46]:
# Fill null values in the Sector column with its mode value
df2['Sector'] = df2['Sector'].fillna(df2['Sector'].mode()[0])


In [47]:
# Check number of unique values in Company_Name column
df2['Company_Name'].nunique()

87

In [48]:
# Check number of unique values in Founders column
df2['Founders'].nunique()

85

In [49]:
# View unique values in Amount column
df2['Amount'].unique()

array(['$6,300,000', '$150,000,000', '$28,000,000', '$30,000,000',
       '$6,000,000', 'Undisclosed', '$1,000,000', '$20,000,000',
       '$275,000,000', '$22,000,000', '$5,000,000', '$140,500',
       '$540,000,000', '$15,000,000', '$182,700', '$12,000,000',
       '$11,000,000', '$15,500,000', '$1,500,000', '$5,500,000',
       '$2,500,000', '$140,000', '$230,000,000', '$49,400,000',
       '$32,000,000', '$26,000,000', '$150,000', '$400,000', '$2,000,000',
       '$100,000,000', '$8,000,000', '$100,000', '$50,000,000',
       '$120,000,000', '$4,000,000', '$6,800,000', '$36,000,000',
       '$5,700,000', '$25,000,000', '$600,000', '$70,000,000',
       '$60,000,000', '$220,000', '$2,800,000', '$2,100,000',
       '$7,000,000', '$311,000,000', '$4,800,000', '$693,000,000',
       '$33,000,000'], dtype=object)

In [50]:
# Function to clean the Amount column

def clean_column(Amount):
    # Remove space,
    Amount = Amount.str.replace(' ', '')

     # Remove comma symbol
    Amount = Amount.str.replace(',', '')
    
    # Remove dollar symbol
    Amount = Amount.str.replace('$', '')

    # Remove single quotation marks surrounding numbers
    Amount = Amount.str.replace(r"'(\d+)'", r'\1')
    
    # Replace 'Undisclosed' with '0'
    Amount = Amount.str.replace('Undisclosed', '0')


    return Amount

# Using the clean_column function
df2['Amount'] = clean_column(df2['Amount'])
df2['Amount'].head(5)


0      6300000
1    150000000
2     28000000
3     30000000
4      6000000
Name: Amount, dtype: object

In [51]:
# Convert the Amount column to float data type
df2['Amount'] = df2['Amount'].astype(float)
df2['Amount'].dtype

dtype('float64')

In [52]:
# Fill null values in the Stage column with its mode value
df2['Stage'] = df2['Stage'].fillna(df2['Stage'].mode()[0])


In [53]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 89 entries, 0 to 88
Data columns (total 10 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Company_Name  89 non-null     object 
 1   Founded       89 non-null     float64
 2   Headquarters  70 non-null     object 
 3   Sector        89 non-null     object 
 4   About         89 non-null     object 
 5   Founders      86 non-null     object 
 6   Investor      89 non-null     object 
 7   Amount        89 non-null     float64
 8   Stage         89 non-null     object 
 9   Year          89 non-null     int64  
dtypes: float64(2), int64(1), object(7)
memory usage: 7.1+ KB


In [54]:
df2.describe().T

count          mean           std     min        25%        50%  \
Founded   89.0  2.014685e+03  2.414895e+00  2004.0     2014.0     2015.0   
Amount    89.0  3.748801e+07  1.038746e+08     0.0  1000000.0  5500000.0   
Year      89.0  2.019000e+03  0.000000e+00  2019.0     2019.0     2019.0   

                75%          max  
Founded      2015.0       2019.0  
Amount   26000000.0  693000000.0  
Year         2019.0       2019.0

In [55]:
df2.duplicated().sum()

0

#### SOURCING THE DATASETS  
The datasets were sourced from a github repository, a onedrive account, and a SQL server database.  
The data at a github repo was read directly using the web url and the dataframe saved as a csv file.  
The data at a onedrive account was downloaded manually due to permission issues.  
The datasets hosted by a SQL server database was queried, and the respective dataframes saved as single files in csv format.  


#### LOADING THE DATASETS  
All four (4) datasets from the three (3) sources were loaded one at a time.  


#### EXPLORING THE DATASETS  
 * 2018 DATASET  
There are 526 records captured by the dataframe.
Some column names were very different from those in the other datasets.
All columns were of object data type which made the descriptive statistics return only count values.
It would seem that there are no null values present in this dataset, but the dash character, '—', exists in the "Industry" and "Amount" columns by 30 and 148 instances respectively.
There were no "Founded" and "Investor" column in the raw dataset.

 * 2019 DATASET
This dataset contained just 89 records.
Even though the column names were closer to those in the 2020 and 2021 datasets, they still need some editing.
Null values exist in five (5) of the nine (9) columns.
All columns were in their appropriate data types except for the "Founded" and "Amount" columns.
The descriptive statistics shows that the minimum value in the "Founded" column is 2004. A start-up should be a company that has been operating for five (5) years or less (a general consensus among experts), so the dataset would be trimmed on that basis.

 * 2020 DATASET
1055 records were found in this dataset.
Of all the columns, only the "Company_Brand" column had no missing values.
The "column10" looks suspiciously under-populated, which could be an indicator of an irrelevant column. This would be investigated during the data cleaning stage.
The descriptive statistics shows that the minimum value for the "Founded" column is 1973. This is definitely too old to be a start-up no matter how you look at it. Once again, the dataset will be trimmed on the basis of their year founded.
Columns with wrong data types will be corrected during the data cleaning stage.

* 2021 DATASET
There are 1209 rows of data in this dataset.
It looks like having the "Founded" and "Amount" columns in the wrong data type was a trend for all the years in review.
The minimum year value looks problematic, like it was in the 2020 dataset.


### DATA CLEANING

2018 DATASET

The "Year" column was created for the 2018 dataset to maintain row uniqueness after merging all datasets into one single dataframe.  
The "Headquarters" and "Sector" columns were created based on the "Location" and "Industry" columns respectively.  
The "Company Name" and "Round/Series" had to be renamed to make them consistent column names in other datasets. The "About Company" was renamed to "About" for simplicity.  
The Location and "Industry" columns were dropped since they will be of no further use to the analytical process.  
The Amount column was in the wrong data type. But before any corrections were made, we looked at the unique values present in the column so that appropriate steps would be taken.  
After viewing the unique values present in the column, we noticed that some records were expressed in terms of rupees, some in terms of dollars, and the rest had no currency designation. We decided to assume the currency of the undesignated records as dollars since there were more records with dollars symbol than those with rupees symbol. The records with the rupees symbol were converted to their respective dollar values in order to have the entire column in terms of dollars only.  
Since the column was in the string data type, we got all non-numeric characters and symbols removed. Then, we converted the entire column into the float data type. Taking a look at the df1 (2018 dataset) dataframe, we can see that there are no null values. The "Amount" column had a mean value of about USD12,626,950 , with minimum and maximum values of USD0.0 and UDS1,000,000,000 respectively. While the "Year" column had its mean, minimum and maximum values as 2018.Duplicate values after cleaning was 1.  
 


#### 2019 DATASET